src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/Checksum.java

```diff
@@ -47,6 +53,12 @@ public static Checksum fromString(KeyType keyType, String hash) throws InvalidCh
     return new Checksum(keyType, HashCode.fromString(hash));
   }
 
+  public static void stringComparison(String hash, String correctHash) throws ICExceptionExtended {
+    if(hash != correctHash) {
```

Comment:
You're using reference equality to compare the strings here. With a few exceptions, such a comparison will always fail. You want to use semantic equality, .equals(). You would need to investigate why there are no failing tests with this code: it's likely this would cause crashes in production, so it would be good to make sure there's test coverage for this.